### PR TITLE
Revert "mul: remove opmath cast sequence (pytorch#9663)"

### PR DESCRIPTION
**Pull request by rajkthakur**
Commit 2a9138a removed `.use_opmathtype_for_compute()` from element-wise 'mul' operation, this breaks mixed-precision accumulation behavior expected by the Neuron compiler that traces/compile on CPU and later execute the binary on neuron hardwares, causing significant accuracy degradation in:

- Llama 3.1 70B models (16.7% throughput drop, accuracy failures)
- Mixtral 8x22B models (accuracy test failures)
- Other transformer models using mixed-precision compilation

Reverts: commit 2a9138a, other changes are result of rebase from r2.9
Fixes: Model accuracy failures with mixed-precision accumulation https://github.com/pytorch/xla/issues/9699

### CODE REVIEW

**Comment by rajkthakur** [comment]
Is the torchax test failure expected?

**Comment by rajkthakur** [comment]
While I understand the suggestion from #8545 to make this backend-specific, I believe a full revert is more appropriate atleast for release branch r2.9:

1. The change affects the fundamental numerical behavior of torch.mul, which is called extensively throughout any model. Even small precision differences compound in:
   • Attention mechanisms with large sequence lengths
   • Gradient accumulation/reduction over many steps and likely XLA based compilers need to explicitly handle this scenario.

2. Risk vs. benefit:
   • Risk:  Our testing shows concrete accuracy regressions and are blocking for production workloads(mixtral, llama) on current compiler, which I guess might impact other users too.
   • Benefit: The original issue #8545 was a question about the seeming unnecessary upcast/downcast that also appears in PyTorch CUDA. The upcast/downcast  can be already removed by using PyTorch autocast so there's no need to fix #8545 for r2.9. We can keep this in master branch for further investigation.

Let me know your thoughts?

**Comment by ysiraichi** [comment]
Thank you for your detailed analysis.

I'm sorry, but I still think that the best way to go about this is to make it backend specific. Actually, now that I'm thinking about it, I think it would make more sense to do it in the lowering step.

That said, since we don't really want to introduce a lot of changes here, and the infrastructure is already in place, I would say that we should make that `use_opmathtype_for_compute()` call backend specific.

It's a simple change (something like [this condition for TPU][1]) that fixes this problem for the Neuron backend, while leaving the other backends with the _"backend-independent"_ choice of not doing that.

> 2. The original issue https://github.com/pytorch/xla/issues/8545 was a question about the seeming unnecessary upcast/downcast that also appears in PyTorch CUDA. The upcast/downcast can be already removed by using PyTorch autocast so there's no need to fix https://github.com/pytorch/xla/issues/8545 for r2.9. We can keep this in master branch for further investigation.

That _"seeming unnecessary upcast/downcast"_ is exactly what we are talking about, here. And, no, PyTorch autocast does not solve this issue (see [this comment][2]).

[1]: https://github.com/pytorch/xla/blob/d0b43ccdc0fa789ace63d66e7b4cdaf7c2904a1a/torch_xla/csrc/aten_xla_type.cpp#L4004
[2]: https://github.com/pytorch/xla/issues/8545#issuecomment-2646410517

File: torch_xla/csrc/aten_xla_type.cpp [hunk 1]
```diff
@@ -2529,13 +2529,26 @@ at::Tensor XLANativeFunctions::mse_loss_backward(const at::Tensor& grad_output,
...
+  auto config =
+      OpConfig([](const XLAInputVector& inputs, at::ScalarType dtype) {
+        return tensor_methods::mul(inputs[0], inputs[1], dtype);
+      })
+          .add_input(self)
+          .add_input(other)
+          .cast_inputs_to_common_dtype();
+
+  // Only use opmathtype for CPU or Neuron backend
+  if (hw_type == XlaDeviceType::CPU || hw_type == XlaDeviceType::NEURON) {
```
Comment (ysiraichi) [inline torch_xla/csrc/aten_xla_type.cpp]: Is this also a problem for CPU?

File: torch_xla/csrc/aten_xla_type.cpp [hunk 2]
```diff
@@ -2529,13 +2529,26 @@ at::Tensor XLANativeFunctions::mse_loss_backward(const at::Tensor& grad_output,
...
-      .cast_inputs_to_common_dtype()
-      .run();
+
+  // Check device type to determine if we need opmathtype for mixed-precision
+  XLA_ASSIGN_OR_THROW(XLATensorPtr xla_self, bridge::GetXlaTensor(self));
+  XlaDeviceType hw_type =
+      static_cast<XlaDeviceType>(xla_self->GetDevice().type());
+
+  auto config =
+      OpConfig([](const XLAInputVector& inputs, at::ScalarType dtype) {
```
Comment (ysiraichi) [inline torch_xla/csrc/aten_xla_type.cpp]: Use the `static_cast` instead of creating a new lambda, here.

File: torch_xla/csrc/aten_xla_type.cpp (hunk 1 above)
Comment (ysiraichi) [inline torch_xla/csrc/aten_xla_type.cpp]: Let's create a `CheckNeuronDevice` function for this.

File: test/test_operations_hlo.py [hunk 3]
```diff
@@ -67,13 +67,13 @@ def test_dropout_by_u8_mask(self):
...
 
-  def test_bfloat16_mul_not_upcast(self):
+  def test_bfloat16_mul_upcast(self):
     a = torch.rand(5, 5, dtype=torch.bfloat16).to('xla')
     b = torch.rand(5, 5, dtype=torch.bfloat16).to('xla')
     c = a * b
     hlo_text = torch_xla._XLAC._get_xla_tensors_hlo([c])
     # Check that the output is not upcasted to float32
-    assert 'f32' not in hlo_text
+    assert 'f32' in hlo_text
```
Comment (ysiraichi) [inline test/test_operations_hlo.py]: Let's keep the test for CPU and TPU as is.

File: torch_xla/csrc/aten_xla_type.cpp (hunk 1 above)
Comment (rajkthakur) [reply torch_xla/csrc/aten_xla_type.cpp]: Yes, Neuron trace on CPU and we have been observing issue with accuracy.

Comment (rajkthakur) [reply torch_xla/csrc/aten_xla_type.cpp]: Will add for CPU

File: test/test_operations_hlo.py (hunk 3 above)
Comment (rajkthakur) [reply test/test_operations_hlo.py]: Will add a condition for TPU

Comment (rajkthakur) [reply test/test_operations_hlo.py]: Change is needed for CPU. This test is run for neuron and cpu and not for tpu. Both should work with my change, let me know if that's fine.

File: torch_xla/csrc/aten_xla_type.cpp (hunk 1 above)
Comment (ysiraichi) [reply torch_xla/csrc/aten_xla_type.cpp]: I don't think I get it. Tracing is done on CPU for all of the devices. `XlaDeviceType` indicates the backend device, where the execution will take place.

So, the question is whether we need this behavior when running computation on CPU.

File: test/test_operations_hlo.py (hunk 3 above)
Comment (ysiraichi) [reply test/test_operations_hlo.py]: I think it's better to have 2 distinct tests: 1 for TPU and CPU (skipping Neuron), and another one for Neuron only (maybe belongs to the Neuron repo?).

File: torch_xla/csrc/aten_xla_type.cpp (hunk 1 above)
Comment (jeffhataws) [reply torch_xla/csrc/aten_xla_type.cpp]: @ysiraichi @qihqi We need the CPU tracing using CPU XLA backend since it is cheaper to do ahead of time tracing/compilation on CPU instance.

See https://awsdocs-neuron.readthedocs-hosted.com/en/latest/frameworks/torch/torch-neuronx/api-reference-guide/inference/api-torch-neuronx-trace.html for the cpu_backend option.

If the PR is not acceptable, are you open to us introducing an environment variable that we can set to add use_opmathtype_for_compute ?

Comment (rajkthakur) [reply torch_xla/csrc/aten_xla_type.cpp]: As per discussion, reverting https://github.com/pytorch/xla/commit/2a9138a26ee257fef05310ad3fecf7c55fe80d73

**Comment by ysiraichi** [comment]
Could you rebase this PR, so that there are only your commits?

**Comment by rajkthakur** [comment]
> Could you rebase this PR, so that there are only your commits?

updated

**Comment by rajkthakur** [comment]
@ysiraichi  CI is completed.

**Comment by jeffhataws** [comment]
Thanks @rajkthakur @ysiraichi . It is merged now and ready to go. @bhavya01 will make a final 2.9 release candidate.
